feat(jest-runtime): expose isTornDown variable

jomendez opened the pull request:
This commit adds a new feature to the Jest runtime package that exposes the `isTornDown` variable as a way to check if the Jest environment has been torn down programmatically. The `isTornDown` variable is a read-only boolean that is set to `true` when the environment is torn down and `false` otherwise.

The commit also adds documentation for the `isTornDown` variable to the Jest runtime module.

Closes #13640

<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md at the root of the project if you have not done so. -->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
The motivation for exposing the isTornDown variable in the Jest runtime package is to provide a way for tests to check if the Jest environment has been torn down programmatically. 
issue https://github.com/facebook/jest/issues/13640 requesting this feature.

## Test plan
### Test Objective
The objective of this test is to verify that the isTornDown variable is correctly exposed in the Jest runtime package and that it can be used to check if the Jest environment has been torn down programmatically.

### Test Steps
1. Clone the Jest repository and make sure that the isTornDown feature has been implemented and is ready for testing.
2. Create a new test file that imports the jest global object and uses the isTornDown variable to check if the Jest environment has been torn down.
3. Run the test file using the jest command.
4. Verify that the test file runs successfully and that the isTornDown variable is correctly set to false when the Jest environment is not torn down.
5. Modify the test file to tear down the Jest environment after the test has run.
6. Run the test file again using the jest command.
7. Verify that the test file runs successfully and that the isTornDown variable is correctly set to true when the Jest environment has been torn down.
8. Modify the test file to check the isTornDown variable in a loop that advances the fake timers by a specified amount of time.
9. Run the test file again using the jest command and verify that the test file stops executing when the Jest environment is torn down.
10. Modify the test file to check the isTornDown variable in a loop that does a lot of work and can time out.
11. Run the test file with a long timeout value (e.g. --timeout=10000) and verify that the test file stops executing when the Jest environment is torn down or the test times out.

### Test Expected Results
- The test file should run successfully and the isTornDown variable should be correctly set to false when the Jest environment is - not torn down.
- The test file should run successfully and the isTornDown variable should be correctly set to true when the Jest environment is torn down.
- The test file should stop executing when the Jest environment is torn down or the test times out.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos if the pull request changes UI. -->

SimenB commented on packages/jest-environment/src/index.ts:
```diff
@@ -172,6 +172,15 @@ export interface Jest {
+   *   return;
+   * }
+   */
+  isTornDown: boolean;
```
this needs to be a function

jomendez replied:
ack

SimenB reviewed (changes requested):
docs seems to be missing? also, please add a test and a changelog entry

mrazauskas commented:
Perhaps it is worth to name it `jest.isEnvironmentTornDown()` instead? See discussion in #13640

SimenB commented:
Yeah, that sounds reasonable. Or even "is test completed"?

jomendez commented:
I messed up on this PR, not sure how it got closed and all the commits diapers, anyways, I created a new PR addressing all the comments here https://github.com/facebook/jest/pull/13741
